---
title: "CatsCradle Spatial Vignette"
author: "Anna Laddach and Michael Shapiro"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{CatsCradleSpatial02}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE, warning = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    fig.dim = c(6,6),
    comment = "#>"
)
```


## Introduction

Here we describe the tools that CatsCradle offers for exploiting the
spatial relationships in spatial transcriptomics data.

 We will
be using a subset  of a Xenium data set that
profiles the mouse hippocampus available from 10x genomics
(https://www.10xgenomics.com/datasets/fresh-frozen-mouse-brain-for-xenium-explorer-demo-1-standard)

Here we visualise this subset coloured by cell type.  ere cell
clusters (Louvain cluster) have been assigned cell type identities
using RCTD along with a reference dataset
(https://www.dropbox.com/s/cuowvm4vrf65pvq/allen_cortex.rds?dl=1)
followed by minimal manual curation. Please note these assignments are
not definitive and are for illustratory purporses only.

```{r}
library(Seurat)
library(CatsCradle)
ImageDimPlot(smallXenium, cols = "polychrome", size = 1)
```

We will want to answer questions like the following:
* Is the expression of particular genes localised?
* Who are the immediate neighbours of a given cell?
* Do certain types of cells tend to co-locate?
  * Is the association of two cell types, (say, glia and macrophages(
  statistically significant?
* For any pair of neighbouring cells, are they engaged
ligand-repceptor interaction?
* Can we classify the different types of tissue neighbourhoods?
* Are different neighbourhood types characterised by
  * The cell types in them?
  * The genes expressed in them?
  * The ligand-receptor interactions taking place in them?


### Neighbourhoods


A key concept here is that of a _neighbourhood_.  A neighbourhood is a
spatially contiguous set of cells.  In every case we will consider, a
neighbourhood is a set of cells _centred on a particular cell_.

The simplest sort of neighbourhood consists of a cell together with
its immediate neighbours.  We compute these as a Delaunay
triangulation using the centroids of the cells:


```{r}
centroids = GetTissueCoordinates(smallXenium)
rownames(centroids) = centroids$cell
delaunayNeighbours = computeNeighboursDelaunay(centroids)
head(delaunayNeighbours)
```

Two edges are neighbours if they appear on the same row.  The
neighbourhood of the cell 16307 consists of the following 7 cells

```{r}
idx = (delaunayNeighbours$nodeA == 16307 |
       delaunayNeighbours$nodeB == 16307)
nbhd = unique(c(delaunayNeighbours$nodeA[idx],
                delaunayNeighbours$nodeB[idx]))
nbhd		
```

We can compute extended neighbourhoods by consider (say) each cells
neighbours, their neighbours, their neighbours and their neighbours.
Here we have such an extend neighbourhood as a list and again as a
data frame with each row given a pair of cells in a common extended
neighbourhood.

```{r}
extendedNeighboursList = getExtendedNBHDs(delaunayNeighbours, 4)
extendedNeighbours = collapseExtendedNBHDs(extendedNeighboursList, 4)
```

Now the extended neighbourhood of  cell 16307 consists of 92 cells.

```{r}
idx = (extendedNeighbours$nodeA == 16307 |
       extendedNeighbours$nodeB == 16307)
nbhd = unique(c(extendedNeighbours$nodeA[idx],
                extendedNeighbours$nodeB[idx]))
length(nbhd)		
```

We can also create neighbours based on Euclidean distance in the tissue.


```{r}
euclideanNeighbours = computeNeighboursEuclidean(centroids,
threshold=20)
```

We see two reasons for looking at these extended neighbourhoods.  One
is that while the Delaunay triangulation might be more appropriate for
looking at cell to cell interactions involving direct contact,
extended neighbourhoods might be more appropriate for studying
interactions based on diffusible ligands.  The second is that when we
go to characterise the _types_ of neighbourhoods, these extended
neighbourhoods exhibit less noise.


We offer two viewpoints as what is "going on" in a neighbourhood.  One
is that a neighbourhood is characterised by the cell types that are
found in it.  In this view, just as a cell expresses genes, a
neighbourhood "expresses" cell types.  A second viewpoint is that a
neighbourhood also expresses genes.  That is, for each neighbourhood,
we can compute the total gene expression across all the cells in that
neighbourhood.

Our exposition will focus on the first viewpoint.  This is not because
we think the first viewpoint is more important or because we
necessarily expect it to be more fruitful.  It's because the second
viewpoint so directly parallels standard Seurat single cell
transcriptomics analyses.  So before we move on to the neighbourhoods
/ cell types viewpoint, we give an overview of the neighbourhoods /
aggregate gene expression viewpoint.

### Neighbourhoods as characterised by gene expression

Here we create a Seurat object based on the aggregate gene expression
in each of our extended neighbourhoods and display the clustering of
the neighbourhoods (called aggregation_clusters) both on the tissue
plot and on the resulting UMAP.

```{r}
agg = aggregateSeuratGeneExpression(smallXenium,extendedNeighbours,
                                    verbose=FALSE)
smallXenium$aggregateNBHDClusters = agg@active.ident
ImageDimPlot(smallXenium,group.by='aggregateNBHDClusters',cols='polychrome')
```

Since neighbourhoods here are indexed by individual cells, the
neighbourhood gene aggregation Seurat object is formally identical to
a standard Seurat object.  In particular, this allows the standard
sorts of analyses including
* Clustering of neighbourhoods into types
* Dimension reduction using PCA, UMAP, tSNE
* Discovery a marker genes for each neighbourhood type
* Plotting of aggregate gene expression on neighbourhood UMAP
* Use of CatsCradle to discover novel clustering of genes based on
their expression across the different neighbourhoods


### Neighbourhoods as characterised by cell types



## Calculation of neighbourhood celltype composition 

Given any of the spatial graphs describing neighbourhoods calculated above,
we can now calculate the cell type composition of neighbourhoods.

```{r}
NBHDByCTMatrix = computeNBHDByCTMatrix(delaunayNeighbours, clusters)
```

In the resulting matrix neighbourhoods are rows and cell types are columns. 
The values in the matrix indicate the number of cells of a given type within a 
neighbourhood. 

Let's do the same for our extended neighbourhoods.
```{r}
NBHDByCTMatrixExtended = 
  computeNBHDByCTMatrix(extendedNeighbours, clusters)
```

## Analysis of contact based interactions between cell types

We can go on to calculate a matrix which gives the fraction of contacts cells of
a given type make with cells of another cell type.
```{r}
cellTypesPerCellTypeMatrix = 
  computeCellTypesPerCellTypeMatrix(NBHDByCTMatrix,clusters)
```

Rows and columns both correspond to cell types, but they are playing
different roles.  For a given row (say, cell type A) the entry in
column B represents the fraction of cells in neighbourhoods of cells
of type A that are of type B.

We can then display this as a force directed graph. Here we choose only to 
display contact based interactions that constitute at least 5% of a cell type's
interactions. Of note, this graph is directed as, for example, 50% of cell type
A's interactions might be with cell type B, but only 5% of cell type B's 
interactions might be with cell type A.   
```{r}
colours = DiscretePalette(length(levels(clusters)), palette = "polychrome")
names(colours) = levels(clusters)

cellTypesPerCellTypeGraphFromCellMatrix(cellTypesPerCellTypeMatrix, 
                                    minWeight = 0.05, colours = colours)
```

Here arrows are directed from rows to columns. Thus, we see an arrow
from 12_Pvalb to 18_pyramidal because neighbourhoods of cells of type
12_Pvalb are composed (across the dataset) of ~15% cells of type
18_pyramidal.  We do not see an arrow from 18_pyramidal to 12_Pvalb
because neighbourhoods of cells of type 18_pyramidal have only 3%
cells of type 12_Pvalb, which falls below the chosen cutoff.

It's worth pointing out the following. The number of edges from cells of
type A to cells of type B is the same as the number of edges from
cells of type B to cells of type A.  Thus the matrix of counts of
these edges is symmetric.  However, numbers of cells of types A and B
are not necessarily equal, and this accounts for the assymmetry in the
fractions. 

```{r}
library(pheatmap)
pheatmap(cellTypesPerCellTypeMatrix)
```



Let's do the same for our expanded nighbourhoods.

```{r}
cellTypesPerCellTypeMatrixExtended = computeCellTypesPerCellTypeMatrix(NBHDByCTMatrixExtended,clusters)

cellTypesPerCellTypeGraphFromCellMatrix(cellTypesPerCellTypeMatrixExtended, minWeight = 0.05, colours = colours)
```
We can also calculate p values (upper tail, one-sided) for whether cell types are more 
commonly neighbours than expected by chance. To do this we compare the actual 
spatial neighbour graph to randomised spatial neighbour graphs where edges are randomised but 
the degree of each vertice is preserved. As this is carried out on the level of 
counts of undirected edges between celltypes the p value matrix is symmetric.

```{r}
cellTypesPerCellTypePValues = computeNeighbourEnrichment(delaunayNeighbours, 
                                                     clusters, verbose = F)
```


Let's plot -log10(pvalue)
```{r}
cellTypesPerCellTypePValuesNegLog = -log10(cellTypesPerCellTypePValues)
pheatmap(cellTypesPerCellTypePValuesNegLog)
```

## Analysis of neighbourhoods based on cell type composition

We can now perform dimensionality reduction and clustering based on 
neighbourhood composition. As the dimensionality of the feature space is 
relatively low (number of cell types) we calculate the UMAP using features
rather than PCs.

```{r}
NBHDByCTSeurat = computeNBHDVsCTSeurat(NBHDByCTMatrix,verbose=FALSE)
```

Add cell type information to the neighbourhoodSeurat object.
```{r}
NBHDByCTSeurat$cellType = clusters
```

Visualise neighbourhood clusters.
```{r}
DimPlot(NBHDByCTSeurat, group.by = c("cellType"), cols = "polychrome", reduction = "umap")
DimPlot(NBHDByCTSeurat, group.by = c("neighbourhood_clusters"), cols = "polychrome", reduction = "umap")
```


We can now add information on neighbourhood clusters to our original xenium object and visualise these on the tissue.

```{r}
smallXenium$NBHDCluster = NBHDByCTSeurat@active.ident
ImageDimPlot(smallXenium, group.by = "NBHDCluster", size = 1, cols = "polychrome")
```

Let's try the same thing with our extended neighbourhoods up to degree 4.
```{r}
NBHDByCTSeuratExtended = computeNBHDVsCTSeurat(NBHDByCTMatrixExtended,
                                               verbose=FALSE)
```
Add cell type information to the NBHDByCTSeuratExtended object.
```{r}
NBHDByCTSeuratExtended$cellType = clusters
```

Visualise extended neighbourhood clusters.
```{r}
DimPlot(NBHDByCTSeuratExtended, group.by = c("cellType"), cols = "polychrome", reduction = "umap")
DimPlot(NBHDByCTSeuratExtended, group.by = c("neighbourhood_clusters"), cols = "polychrome", reduction = "umap")
```
We can now add information on neighbourhood clusters to our original Xenium object and visualise these on the tissue.
```{r}
smallXenium$NBHDClusterExtended= 
  NBHDByCTSeuratExtended@active.ident
ImageDimPlot(smallXenium, group.by = c("NBHDClusterExtended"), 
             size = 1, cols = "polychrome")
```


Here we retrieve fewer clusters, and these describe tissue architecture 
rather than small variations in cellular niche.  

We leave it to the user to decide which approach is most applicable to the 
biological question at hand.


## Relating cell type clusters to neighbourhood clusters

We can now ask how clusters defined transcriptomically (cell type)
relate to those defined based on neighbourhoods (cell niche).  Put
differently, each cell has its (transcriptomic) cell type and sits at
the center of a neighbourhood with a given neighbourhood type.  For
each cell type A and each neighbourhood type B, we can ask what
percentage of the time a cell of type A sits at the center of a
neighbourhood of type B.

```{r}
CTByNBHDCluster = table(NBHDByCTSeurat$cellType,NBHDByCTSeurat@active.ident)
CTByNBHDCluster = CTByNBHDCluster/rowSums(CTByNBHDCluster)

rownames(CTByNBHDCluster) = paste0("CellType",rownames(CTByNBHDCluster))
colnames(CTByNBHDCluster) = paste0("NBHDCluster",colnames(CTByNBHDCluster))

pheatmap(CTByNBHDCluster,
      fontsize_row=8,
      fontsize_col=8,
      cellheight=10,
      cellwidth=10)

sankeyFromMatrix(CTByNBHDCluster)
```

This allows us to see which cell types share the same niche (neighbourhood 
clusters). Let's also perform this analysis using the extended neighbourhoods. 


```{r}
CTByNBHDClusterExtended = table(NBHDByCTSeuratExtended$cellType,NBHDByCTSeuratExtended@active.ident)
CTByNBHDClusterExtended = CTByNBHDClusterExtended/rowSums(CTByNBHDClusterExtended)

rownames(CTByNBHDClusterExtended) = paste0("CellType",rownames(CTByNBHDClusterExtended))
colnames(CTByNBHDClusterExtended) = paste0("NBHDCluster",colnames(CTByNBHDClusterExtended))

pheatmap(CTByNBHDClusterExtended,
      fontsize_row=8,
      fontsize_col=8,
      cellheight=10,
      cellwidth=10)

sankeyFromMatrix(CTByNBHDClusterExtended)
```

## Analysing cell types based on their neighbourhoods

Now we perform dimensionality and clustering of cell types, based on the 
neighbourhoods they are found in. Note that this is based on the transpose of 
the matrix used in the previous section.

First we create a seurat object for a cell type by neighbourhood matrix (the 
transposed NBHDByCTMatrix). Here we have a 
large feature space (numbers of cells) however low number of observations 
(number of cell types). Therefore we compute the UMAP using PCs, however due to 
the low number of observations we need to set a lower value for n.neighbours.


```{r}
CTByNBHDSeurat = 
  computeNBHDVsCTSeurat(t(NBHDByCTMatrix), npcs = 10, 
                        transpose = T, resolution = 1, n.neighbors = 5,
			verbose=FALSE)

CTByNBHDSeurat$cellType = colnames(CTByNBHDSeurat)

DimPlot(CTByNBHDSeurat, group.by = "cellType", cols = "polychrome", 
        reduction = "umap", label = T)
```






We can also compute a graph embedding.
```{r}
CTByNBHDSeurat= computeGraphEmbedding(CTByNBHDSeurat)

DimPlot(CTByNBHDSeurat,group.by = "cellType", cols = "alphabet", reduction = "graph", label = T)
```

We achieve only one cluster using the Louvain algorithm. Let's try hierachical 
clustering.
```{r}
pca = Embeddings(CTByNBHDSeurat, reduction = "pca")
res = pheatmap(pca)
```


Here we decide to cut the tree at the correct level to form 11 clusters.
```{r}
CTClust = cutree(res$tree_row, k = 11)
CTByNBHDSeurat$neighbourhood_clusters = factor(CTClust)
```

Let's investigate how neighbourhood clusters relate to cell type clusters.
```{r}
averageExpMatrix = getAverageExpressionMatrix(NBHDByCTSeurat,
                           CTByNBHDSeurat,
			    clusteringName='neighbourhood_clusters')
averageExpMatrix = tagRowAndColNames(averageExpMatrix,
                                     ccTag='neighbourhoodClusters_',
                                     gcTag='cellTypeClusters_')
pheatmap(averageExpMatrix,
      fontsize_row=8,
      fontsize_col=8,
      cellheight=10,
      cellwidth=10)


sankeyFromMatrix(averageExpMatrix)
```

## Detection of genes with spatially variable expression.

Given a neighbour graph describing the spatial relationships between cells, 
we can compute spatial autocorrelation (here we use Moran's I). This describes
how clustered gene expression values are in space. We derive an upper tail p 
value based on simulations where expression values for genes are permuted. N.B 
in this implementation equal weights are given to all neighbours of a cell.

```{r}
moransI = runMoransI(smallXenium, delaunayNeighbours, assay = "SCT", 
                     layer = "data", nSim = 100, verbose = FALSE)
```
Look at most spatially autocorrelated genes.
```{r}
head(moransI)
```

Look at least spatially autocorrelated genes.
```{r}
tail(moransI)
```

Visualise most spatially autocorrelated gene.
```{r}
ImageFeaturePlot(smallXenium, "Nwd2")
```
We can compute a p-value for the localisation of gene expression with the
following

```{r eval = FALSE}
## We need the coordinates as a matrix:
cellCoords = data.matrix(centroids[,1:2])
Nwd2 = FetchData(smallXenium,'Nwd2')
## We need to threshold gene expression:
Nwd2 = Nwd2[,1] > 0.01
pValue = medianComplementPValue(coords,Nwd2)
```

This returns a p-value of 0.001, here limited by the default number of
permutation trials used by medianComplementPValue()


Visualise least spatially autocorrelated gene.
```{r}
ImageFeaturePlot(smallXenium, "Trbc2")
```

## Ligand receptor analysis

We can perform an analysis of ligand receptor interactions in order to infer 
communication between cells. 
Here we calculate whether interactions between ligand receptor pairs occur on 
edges of the spatial graph, i.e. between neighbouring cells, where cell A 
expresses a ligand and cell B expresses a receptor. Note that the distance 
between cell A and cell B will vary depending on how the graph has been 
constructed, larger distances may be desired for analyses involving diffusable 
ligands and receptors. We also analyse analyse the number of ligand receptor 
interactions that occur between cells from the same cluster and between cells 
from different clusters. Note that clusters may represent cell type, or another 
property of the cells such as their neighbourhood type. We leave this for the 
user to decide. We then calculate how frequently the number of ligand-receptor 
interactions observed between/within clusters is higher than in simulated data,
this allows us to assign a pvalue for the enrichment of these interactions.

Perform interaction analysis:
```{r}
ligandReceptorResults = performLigandReceptorAnalysis(smallXenium, delaunayNeighbours, 
                                                "mouse", clusters,verbose=FALSE)
```


Look at interactions on edges:
```{r}
head(ligandReceptorResults$interactionsOnEdges[,1:10],10)
```

Look at total interactions between/within clusters:
```{r}
head(ligandReceptorResults$totalInteractionsByCluster[,1:10],10)
```

Look at mean interactions per edge between/within clusters:
```{r}
head(ligandReceptorResults$meanInteractionsByCluster[,1:10],10)
```

Look at number of times observed interactions are more frequent than expected:
```{r}
head(ligandReceptorResults$simResults[,1:10],10)
```

Look at pvalues for the enrichment of interactions.
```{r}
head(ligandReceptorResults$pValues,10)
```


Plot a heatmap showing -log10(pvalues) for the enrichment of ligand-receptor 
interactions between pairs of clusters. 

```{r}
ligRecMatrix = makeLRInteractionHeatmap(ligandReceptorResults, clusters, colours = colours, labelClusterPairs = F)
```



```{r}
M = makeLRInteractionHeatmap(ligandReceptorResults, clusters, labelClusterPairs = F)
```


```{r}
cellTypePerCellTypeLigRecMatrix = makeSummedLRInteractionHeatmap(ligandReceptorResults, clusters, "total")
```

```{r}
cellTypePerCellTypeLigRecMatrix = makeSummedLRInteractionHeatmap(ligandReceptorResults, clusters, "mean", logScale = T)
```
```{r}
hist(cellTypePerCellTypeLigRecMatrix)
```
Visualise in graph format.
```{r}
cellTypesPerCellTypeGraphFromCellMatrix(cellTypePerCellTypeLigRecMatrix/5, 
                                    minWeight = 0.08, colours = colours)
```


Create a Seurat object where each point represents an edge between cells, and spatial coordinates are thecentroids of edges between cells. The "expression matrix" is the binarised presence/absence of an interaction (ligand receptor pair) on an edge. This is useful for visualising where ligand receptor interactions occur spatially.
```{r}
edgeSeurat = computeEdgeSeurat(ligandReceptorResults, centroids)
```


Visualise Penk-Htr1f interaction.
```{r}
ImageFeaturePlot(edgeSeurat, features = "Penk-Htr1f")
```

We can compute a spatial graph where edges in the original delaunayNeighbours 
become nodes and A-B edges (in the original graph) become connected to all 
A- edges and all B- edges. This allows us to perfom graph-based associated with 
the spatial localisation of ligand receptor pairs on edges.  
```{r}
edgeNeighbours = computeEdgeGraph(delaunayNeighbours)
```


Compute Moran's I for the spatial autocorrelation of ligand-receptor interactions.
```{r}
moransILigandReceptor = runMoransI(edgeSeurat, edgeNeighbours, assay = "RNA", 
                     layer = "counts", nSim = 100)
```

View most spatially autocorrelated ligand-receptor interactions.
```{r}
head(moransILigRec)
```

View least spatially autocorrelated ligand-receptor interactions.
```{r}
tail(moransILigRec)
```


View ligand-receptor interaction with the highest spatial autocorrelation (Moran's I).
```{r}
ImageFeaturePlot(edgeSeurat, "Penk-Htr1f")
```

View ligand-receptor interaction with the lowest spatial autocorrelation (Moran's I).
```{r}
ImageFeaturePlot(edgeSeurat, "Sst-Gpr17")
```

